### PR TITLE
[react] add forwardRef and createRef functions for React v16.3

### PR DESCRIPTION
**Pull request by ferdaber**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  - [forwardRef](https://github.com/facebook/react/blob/master/packages/react/src/forwardRef.js)
  - [createRef](https://github.com/facebook/react/blob/master/packages/react/src/ReactCreateRef.js)
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

**Comment by jgoz** [comment]
@ferdaber The [build failures](https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/360136770#L3248-L3268) here look related to your changes.

```
...
Type 'RefObject<{}>' is not assignable to type 'RefObject<ClassicComponent<Props, ComponentState>>'.
```

**Comment by ferdaber** [comment]
You're right, I'll fix it when I get back home.

File: types/react/index.d.ts
Comment (Kovensky) [inline types/react/index.d.ts]: Is it actually a component, or just a callback?

Though arguably any function that returns `ReactNode` is a component if it reads its parameters in a compatible way.

File: types/react/index.d.ts
Comment (Kovensky) [inline types/react/index.d.ts]: Would it be possible to use a secret `unique symbol` to brand this interface and make it only constructible via `createRef`? Or would any `unique symbol`s we declare be visible to user code, even if it doesn't exist?

File: types/react/index.d.ts
Comment (ferdaber) [reply types/react/index.d.ts]: The source code even calls it `render` as its parameter name, I called it a component here to make it look and feel more familiar to userland, as I think a lot of use cases for this would be inlining a functional component like:

```tsx
const MyComponent = React.forwardRef(
  function MyComponent(props: {}, ref: React.Ref<HTMLDivElement>) {
    return <div ref={ref} />
})
```

and having the interface have the word `Component` in it will give a bit of extra guidance as to how the user should structure the callback.

File: types/react/index.d.ts
Comment (ferdaber) [reply types/react/index.d.ts]: I'm not sure what you mean -- do you mean we should add another property in the interface to ensure that users don't try to pass any old object into `ref`?

I tried something like this, which does work, let me know if that works for you:
```tsx
// top level declaration
declare const ReactRef: unique symbol;

declare namespace React {
  // should only be constructed by createRef
  interface RefObject<T> {
    readonly current: T | null;
    [ReactRef]: T
}
```

```tsx
// ReactRef is unable to be imported since it's a non-exported const
import { RefObject, Component, createRef } from 'react'

class MyComponent extends Component {
    mySneakyRef: RefObject<HTMLDivElement> = { // $ExpectError
        current: null
    }
    myTrueRef: RefObject<HTMLDivElement> = createRef()

    render() {
        return (
            <>
                <div ref={this.mySneakyRef} />    
                <div ref={this.myTrueRef} />    
            </>    
        )
    }
}
```

Comment (ferdaber) [reply types/react/index.d.ts]: Interestingly though, the source really does just construct a plain object and return it (despite the fact that it's sealed in `DEV` mode), so a user passing in a plain object into `ref` will also find that it works:

https://github.com/facebook/react/blob/master/packages/react/src/ReactCreateRef.js

File: types/react/test/index.ts
Comment (pelotom) [inline types/react/test/index.ts]: This type annotation can just be removed entirely:

```ts
const props = {
  // ...
```

File: types/react/index.d.ts
Comment (pelotom) [inline types/react/index.d.ts]: My guess is it will be more annoying than useful to treat `current` as nullable (even though it technically is).

Comment (Kovensky) [reply types/react/index.d.ts]: This is the same behavior as the current argument to ref callbacks; you also can brute force your way through it by using a `!`.

File: types/react/index.d.ts
Comment (Kovensky) [reply types/react/index.d.ts]: That's interesting 🤔 

I thought it'd brand the returned object, similar to the context/provider pair, but I guess they get away with not branding it since it's not something you give as the first argument to `createElement` 😄

File: types/react/index.d.ts
Comment (pelotom) [reply types/react/index.d.ts]: Yes, but in 2.7+ you can use definite assignment assertion to pretend the ref is non-null:

```ts
class Foo extends React.Component {
  elem!: HTMLInputElement

  render() {
    return <input ref={elem => elem && (this.elem = elem)} />;
  }

  componentDidMount() {
    this.elem.focus();
  }
}
```

No such recourse will exist with `current` if it's nullable. Also, if people want to opt in to a nullable ref they can do so:

```ts
class Foo extends React.Component {
  ref = React.createRef<HTMLInputElement | null>();

  render() {
    return <input ref={this.ref} />;
  }
}
```

But if it's required to be nullable in the definition, they are unable to opt out.

Comment (Kovensky) [reply types/react/index.d.ts]: The definite assignment is a different problem. You are checking if `elem` is `null` or not with `elem && (this.elem = elem)`, and so you cannot assign the `null` value to it. The definite assignment lets you avoid the `undefined` state, but `null` is a normal part of the API.

Comment (pelotom) [reply types/react/index.d.ts]: What it allows you to do is treat `this.elem` as non-null everywhere _else, i.e. to avoid having to use `this.elem!` everywhere.

Comment (Kovensky) [reply types/react/index.d.ts]: That behavior is only correct if your subtree is always mounted. If your subtree is mounted conditionally, you now have a hidden `null` reference. The `| null` here is _correct behavior_.

Comment (pelotom) [reply types/react/index.d.ts]: Yes, and if you happen to know it’s always mounted this is exactly the use case for the definite assignment assertion feature. It’s a small compromise in type safety but very convenient for situations where you know better than the compiler. This is why the feature exists. By forcing `current` to be `| null` we lose the option to exercise such discretion, so it’s a step backwards. Again, you can always opt in to perfect type safety by declaring the type parameter of the `RefObject` to be `| null`, but you cannot opt out if the decision is made for you.

Comment (Kovensky) [reply types/react/index.d.ts]: No, this is _not_ the use case for definite assignment assertion. The definite assignment is for avoiding the presence of `undefined` _specifically_.

Do not confuse `null` with `undefined`.

Comment (pelotom) [reply types/react/index.d.ts]: Hm, you may be right, but how is it a potential memory leak? Doesn't React release instances once unmounted, thereby releasing their fields?

**Comment by mhegazy** [comment]
@Kovensky and @pelotom what is the final call on this one, can i get a signoff or request changes mark from you on this PR.

**Review by Jessidhia — APPROVED** [review]

**Comment by MichaelTontchev** [comment]
In what version of TS can we expect this?

**Comment by ferdaber** [comment]
@MichaelTontchev it should be usable by TS v2.6+, I had thought you can just `npm i @types/react@^16.3`, not sure how releases are done once PRs are merged.

**Comment by smashraid** [comment]
Can someone help me with this? I would like to know the correct way to use. Focus method doesn't exist 

`"@types/react": "^16.4.6",`
`"@types/react-dom": "^16.0.6",`
`"react": "^16.4.1",`
`"react-dom": "^16.4.1"`

image

**Comment by ferdaber** [comment]
@smashraid that is correct, since the ref object isn't always guaranteed to be populated with a `current`. You'd need to either null check it before calling focus or assert that it's always populated:
```ts
// if you're sure that the ref is always there (no conditional rendering, for example)
this.step.current!.focus();
// otherwise null check it
this.step.current && this.step.current.focus();
```

**Comment by smashraid** [comment]
Hi, can someone help me? I don't know if this is the correct implementation of **React.forwardRef**
image
image
image

Then I guess I can access to the component ref like this
`this.panel.current!.style.backgroundColor = "red";`

**Pull request merged**
